File: core/src/main/java/net/year4000/mapnodes/game/regions/PointRegion.java
/*
 * Copyright 2016 Year4000. All Rights Reserved.
 */

package net.year4000.mapnodes.game.regions;

import com.flowpowered.math.vector.Vector3i;
import com.google.common.collect.ImmutableSet;
import net.year4000.utilities.Conditions;
import net.year4000.utilities.Utils;

import java.util.Optional;
import java.util.Set;

/** A region that contains a single point */
public final class PointRegion extends AbstractRegion {
    private Vector3i point;

    /** Allow creation of a PointRegion via x,y,z cords */
    public PointRegion(int x, int y, int z) {
        this(new Vector3i(x, y, z));
    }

    /** Create the region with vector3i point */
    public PointRegion(Vector3i point) {
        this.point = Conditions.nonNull(point, "point");
    }

    @Override
    public boolean contains(Vector3i vector3i) {
        return point.equals(vector3i);
    }

    @Override
    public Optional<Set<Vector3i>> getPoints() {
        return Optional.of(ImmutableSet.of(point));
    }

    @Override
    public boolean equals(Object other) {
        return Utils.equals(this, other);
    }

    @Override
    public int hashCode() {
        return Utils.hashCode(this);
    }

    @Override
    public String toString() {
        return Utils.toString(this);
    }
}

File: core/src/main/js/net/year4000/mapnodes/js/regions/point_region.js
/*
 * Copyright 2017 Year4000. All Rights Reserved.
 */

/** A region that contains a single point, must wrap with THREE.Vector3 */
class PointRegion extends AbstractRegion {

  constructor(x, y, z) {
    super()
    this._point = new THREE.Vector3(x, y, z)
  }

  /** Get the point of this region */
  get points() {
    return Immutable.Set.of(this._point)
  }

  /** Get the x cord */
  get x() {
    return this._point.x
  }

  /** Get the y cord */
  get y() {
    return this._point.y
  }

  /** Get the z cord */
  get z() {
     return this._point.z
   }

  /** Create a copy of the point for vector3 */
  clone() {
    return this._point.clone()
  }

  // Override
  valueOf() {
    // todo support for Immutable Sets
    return (this.x * 211 + this.y) * 97 + this.z
  }
}
